### PR TITLE
Add plugins

### PR DESCRIPTION
**Pull request by joehoyle**
Added SAML 2.0 from https://github.com/humanmade/wp-simple-saml which from what I can see looks pretty good (cc @shadyvb !)

I also added "WordPress" via https://github.com/humanmade/delegated-oauth2, as it's pretty easy and I think it's going to be useful to do platform -> platform stack SSO too.

### CODE REVIEW

File: docs/saml-2-0.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+The SSO module includes support for SAML 2.0 as a Service Provider. To enable SAML 2.0, you must define the `saml` setting in the Platform configuration. You'll need a copy of your SAML IdP Metadata XML too.
+
+To enabled SAML 2.0 support, define the following options in the configuration file, and provider the SAML IdP Metadata in a file:
+
+
+```json
+"sso": {
+	"saml": {
+		required: true | false
```
Comment (roborourke) [inline docs/saml-2-0.md]:
```suggestion
		"required": true | false
```

File: docs/saml-2-0.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# SAML 2.0
+
+The SSO module includes support for SAML 2.0 as a Service Provider. To enable SAML 2.0, you must define the `saml` setting in the Platform configuration. You'll need a copy of your SAML IdP Metadata XML too.
+
+To enabled SAML 2.0 support, define the following options in the configuration file, and provider the SAML IdP Metadata in a file:
```
Comment (roborourke) [inline docs/saml-2-0.md]:
```suggestion
To enabled SAML 2.0 support, define the following options in the configuration file, and provide the SAML IdP Metadata in a file:
```

File: docs/saml-2-0.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+Contents of `config/sso/saml-idp-metadata.xml`:
+
+```xml
+<?xml version="1.0"?>
+<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" entityID="http://localhost:8082/simplesaml/saml2/idp/metadata.php">
+  <md:IDPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
+    <md:KeyDescriptor use="signing">
+      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
+        <ds:X509Data>
+          <ds:X509Certificate>//redactoed</ds:X509Certificate>
```
Comment (roborourke) [inline docs/saml-2-0.md]:
```suggestion
          <ds:X509Certificate>//redacted</ds:X509Certificate>
```

File: docs/saml-2-0.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        </ds:X509Data>
+      </ds:KeyInfo>
+    </md:KeyDescriptor>
+    <md:KeyDescriptor use="encryption">
+      <ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
+        <ds:X509Data>
+          <ds:X509Certificate>//redactoed</ds:X509Certificate>
```
Comment (roborourke) [inline docs/saml-2-0.md]:
```suggestion
          <ds:X509Certificate>//redacted</ds:X509Certificate>
```

File: docs/saml-2-0.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        </ds:X509Data>
+      </ds:KeyInfo>
+    </md:KeyDescriptor>
+    <md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="http://localhost:8082/simplesaml/saml2/idp/SingleLogoutService.php"/>
```
Comment (roborourke) [inline docs/saml-2-0.md]: Where do this localhost URL come from? Can we use an SSL URL? Guessing it's fine as within the network but might be a question that comes up

File: docs/wordpress.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+To enabled WordPress SSO, you must create an OAuth 2 application on the WordPress site, and set the Callback URL on the OAuth client to `$site_url/hm-delegated-auth-callback`, for example `https://my-site.altis.dev/hm-delegated-auth-callback`. Once you have published and approved the OAuth application on your WordPress site, take note of the Public Key for the client.
+
+Provide the REST BASE of your WordPress site (usually `https://examples.com/wp-json/`), along with the OAuth client id (public key) obtained in the previous step. This is done via the `sso.wordpress` setting option:
+
+```json
+"sso": {
+	"wordpress": {
+		"server-rest-base": "https://examples.com/wp-json/",
+		"oauth2-client-id: "XXXXXXXXXXX"
```
Comment (roborourke) [inline docs/wordpress.md]:
```suggestion
		"oauth2-client-id": "XXXXXXXXXXX"
```

File: docs/wordpress.md [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# WordPress
+
+The SSO module allows using a WordPress site as the authentication provider for the Platform site. This works via OAuth 2.0, as such the WordPress site you want to use as the authentication provider must be using the [OAuth 2.0 WordPress Plugin](https://github.com/WP-API/OAuth2).
+
+To enabled WordPress SSO, you must create an OAuth 2 application on the WordPress site, and set the Callback URL on the OAuth client to `$site_url/hm-delegated-auth-callback`, for example `https://my-site.altis.dev/hm-delegated-auth-callback`. Once you have published and approved the OAuth application on your WordPress site, take note of the Public Key for the client.
```
Comment (roborourke) [inline docs/wordpress.md]: Is this not a default we can infer?

File: load.php [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace HM\Platform\SSO;
+
+use function HM\Platform\get_config;
+use function HM\Platform\register_module;
+
+require_once __DIR__ . '/inc/saml/namespace.php';
```
Comment (roborourke) [inline load.php]: curious about this approach - any reason to load these by default and not just based on whether they're enabled?

File: docs/saml-2-0.md (hunk 5 above)
Comment (joehoyle) [reply docs/saml-2-0.md]: This is just an example metadata file, so it's not something _we_ provide, it's something they would get from their sysadmins. Or are you just thinking it's bad to have _any_ example data that is for http?

File: docs/wordpress.md (hunk 7 above)
Comment (joehoyle) [reply docs/wordpress.md]: This is about configuring oauth2 on your WordPress site you want to be the provider, not the Platform site, so we don't control anything "over there"

File: load.php (hunk 8 above)
Comment (joehoyle) [reply load.php]: I think in discussions with @rmccue etc, we decided it was simpler to include all code ahead of time, regardless if used. The exception to that is side-effects inclusions, like wordpress plugin files

**Comment by joehoyle** [comment]
@roborourke looking at https://github.com/clifgriffin/simple-ldap-login/blob/master/Simple-LDAP-Login.php, I think we might need a strong code review and possible clean-up if we want' to include that. We have LDAP slated post-release atm, so I'd say we push ahead with what we have here. Looks like it shouldn't be too hard to get that plugin up to scratch though.

File: load.php (hunk 8 above)
Comment (rmccue) [reply load.php]: Yeah, I think if there's no side effects, we should always include the code unconditionally. It slightly improves the opcache performance (although a micro-optimisation), and simplifies the code flow.

File: docs/saml-2-0.md (hunk 5 above)
Comment (roborourke) [reply docs/saml-2-0.md]: Right - I got that much that it was example data, more that `http://localhost:8082` seemed very specific so thought it was a default. If it's not something that matters there particularly then all good, just wanted to be sure.

File: load.php (hunk 8 above)
Comment (roborourke) [reply load.php]: Ok, I need to update some things in the SEO module to reflect that then 👍

**Comment by roborourke** [comment]
Cool cool cool 😎 lgtm

**Pull request merged**